### PR TITLE
Maintain dialogue history: as a utility function which any model can use

### PR DESCRIPTION
**Pull request by jaseweston**

### CODE REVIEW

File: parlai/core/utils.py [hunk 1]
```diff
@@ -185,3 +185,50 @@ def sort_data(data, key='text_label', method='spaces'):
...
+                            historyLength=1, useReplies="labels",
+                            dict=None, useStartEndIndices=True):
+    """Keeps track of dialog history, up to a truncation length.
+    Either includes replies from the labels, model, or not all using param 'replies'."""
+
+    def parse(txt):
+        if dict is not None:
+            vec =  dict.txt2vec(txt)
+            if useStartEndIndices:
+                parsed_x = deque([dict[dict.start_token]])
```
Comment (alexholdenmiller) [inline parlai/core/utils.py]: maxlen=?

Comment (alexholdenmiller) [reply parlai/core/utils.py]: resolved, this is right!

File: parlai/core/utils.py [hunk 2]
```diff
@@ -185,3 +185,50 @@ def sort_data(data, key='text_label', method='spaces'):
...
+
+    if history['episode_done']:
+        history['dialog'].clear()
+        history['episode_done'] = False
+
+    if useReplies != 'none':
+        if useReplies == 'model':
+            if reply != '':
+                history['dialog'].extend(parse(reply))
+            r = reply
```
Comment (alexholdenmiller) [inline parlai/core/utils.py]: looks like this `r` isn't used

File: parlai/core/utils.py [hunk 3]
```diff
@@ -185,3 +185,50 @@ def sort_data(data, key='text_label', method='spaces'):
...
+        else:
```
Comment (alexholdenmiller) [inline parlai/core/utils.py]: elif

File: parlai/core/utils.py [hunk 4]
```diff
@@ -185,3 +185,50 @@ def sort_data(data, key='text_label', method='spaces'):
...
+            if 'labels' in observation:
```
Comment (alexholdenmiller) [inline parlai/core/utils.py]: elif

**Comment by jaseweston** [comment]
ok, fixed

**Pull request merged**